### ISSUE TITLE
Support for PROTECT on foreign keys

### ISSUE THREAD

**Issue by jorge-marques** [torchbox/wagtail#1602]
Are there plans to support `on_delete=PROTECT` in the admin? Right now ProtectedError is not caught.

**Comment by gasman** [comment]
There are no current plans for it - we've found that on_delete=CASCADE and on_delete=SET_NULL are sufficient for our needs. If you think this would be a useful addition, we'd happily consider a pull request for it, though.

**Comment by tomdyson** [comment]
We're experiencing this (failure to catch `ProtectedError`) on client projects.

**Comment by prydie** [comment]
:+1:

**Comment by ababic** [comment]
We were experiencing this recently too, so I added some handling for it into wagtailmodeladmin's delete view. However, It's only triggered after user confirmation, when the failed delete raises the exception. I suppose it'd be more helpful to flag it up before that and not even allow an attempt to be made. 

For me, that idea opens up a can of worms, though. If I were to provide up-front feedback about protected relationships, surely it's just as (if not, more) important to flag up ones using CASCADE also? And what about  relationships using SET_NULL? should they at least get a mention? What if the instance has relationships with various other objects using all of those methods? It was all bit much for me to tackle alone.

**Comment by pacotole** [comment]
I made a PR to catch ProtectedError on image delete.
This is the link if it cant help you https://github.com/torchbox/wagtail/pull/3082

**Comment by hakjoon** [comment]
Is this issue still accurate given the `wagtailcore.W001` check which recommends using `models.PROTECT` and to explicitly not use `CASCADE`? [wagtailcore.W001](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/models.py#L558)

**Comment by gasman** [comment]
@hakjoon Yes, it's still relevant. The warning suggests using PROTECT or SET_NULL; PROTECT is definitely a better choice than CASCADE (because a page deleted through CASCADE will corrupt the tree) but still isn't ideal because it throws an uncaught exception. (Also, the warning only applies to pages; the issue here is relevant to snippet / modeladmin models too.)

**Comment by hakjoon** [comment]
@gasman Sorry I think I framed my question poorly.  I was wondering if the current position is that there are no plans to support `PROTECT` given the message `wagtailcore.W001` presents.  Also I was wondering about the statement that `CASCADE` is sufficient, given the messaging that `CASCADE` should not be used.

We set a OneToOne field to `PROTECT` due to the warning but that generated an exception in ModelCluster that `PROTECT` was not supported, and then came across this issue which said there no plans to support `PROTECT`.

I was confused by the apparent mixed messaging which seems to just be an issue of things being in transition. Thanks for the clarification.

**Comment by Toshakins** [comment]
Hi, everybody. Somehow decided to fix this issue, created a pull request: https://github.com/wagtail/wagtail/pull/3525

Not sure if somebody has seen it. I understand that maybe code needs polishing or the issue has low priority - just wanted to remind that any feedback is welcome.

**Comment by BertrandBordage** [comment]
If the aim of Wagtail’s whole admin is to use ModelAdmin everywhere, then the solution can be pretty straightforward, it’s simply a permissions question.

Modifying [this method](https://github.com/wagtail/wagtail/blob/37bbbb9dbabefc9a658cf59eb2311b075f0b7b2e/wagtail/contrib/modeladmin/helpers/permission.py#L85-L91) would fix the issue.

It could be something like:
```python
from django.db.models.deletions import Collector, ProtectedError

[…]

def user_can_delete_obj(self, user, obj):
    try:
        Collector(obj._state.db).collect([obj])
    except ProtectedError:
        return False
    perm_codename = self.get_perm_codename('delete')
    return self.user_has_specific_permission(user, perm_codename)
```

**Comment by lb-** [comment]
Update - a potential sub-case of this issue has been resolved via this PR
https://github.com/wagtail/wagtail/pull/5437

**Comment by pySilver** [comment]
If you want a user friendly error instead of 500 server error, consider using this hook:

```python
@hooks.register("before_delete_page")
def on_page_delete(request, page):
    if request.method != "POST":
        return

    try:
        next_url = get_valid_next_url_from_request(request)
        parent_id = page.get_parent().id
        page.delete()

        messages.success(
            request, _("Page '{0}' deleted.").format(page.get_admin_display_title())
        )

        for fn in hooks.get_hooks("after_delete_page"):
            result = fn(request, page)
            if hasattr(result, "status_code"):
                return result

        if next_url:
            return redirect(next_url)
        return redirect("wagtailadmin_explore", parent_id)

    except ProtectedError as exc:
        protected_objects = {}
        for obj in exc.protected_objects.all():
            model_name = obj._meta.verbose_name_plural
            if model_name in protected_objects:
                protected_objects[model_name].append(str(obj))
            else:
                protected_objects[model_name] = [str(obj)]

        dependency_summary = []
        for model_name, items in protected_objects.items():
            if len(items) == 1:
                items_summary = _("{0} ({1})").format(model_name, items[0])
            else:
                items_summary = _("{0} ({1} and {2} more...)").format(
                    model_name, items[0], len(items) - 1
                )

            dependency_summary.append(items_summary)

        messages.error(
            request,
            _("Page '{0}' cannot be deleted while it's used by {1}").format(
                page.get_admin_display_title(), ", ".join(dependency_summary)
            ),
        )
        return redirect("wagtailadmin_pages:delete", page.id)
```

**Comment by spapas** [comment]
Unfortunately the solution @pySilver posted works only for pages (that have the `before_delete_page` hook). It isn't applicable when deleting images, documents or whatever else that is used needs to be deleted.

What I propose instead is to use the following middleware:

```
class HandleProtectionErrorMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def process_exception(self, request, exception):
        from django.db.models.deletion import ProtectedError
        from django.contrib import messages
        from django.http import HttpResponseRedirect

        if isinstance(exception, ProtectedError):
            messages.error(
                request,
                "The object you are trying to delete is used somewhere. Please remove any usages and try again!.",
            )
            return HttpResponseRedirect(request.path)

        return None

    def __call__(self, request):
        response = self.get_response(request)
        return response
```

Just add it to the end of your `MIDDLEWARE` list. If a `ProtectedError` is thrown it will catch it and add a proper message to the view gracefully handling the excception. I think that this middleware *needs* to be a part of the wagtail default installation or at least mentioned somewhere in the docs so that people can use it if they want it.

**Comment by pySilver** [comment]
Well my case was about pages becase.. that was my case :)

I'm strongly against middleware that runs on every request. It should be solved somewhere else.

**Comment by spapas** [comment]
@pySilver I don't think there's a possible *DRY* way of resolving that issue without using a middleware :) Also it's just a recommendation, if somebody wants to use knowing the consequences (which shouldn't be that important) why not use it ?

**Comment by pySilver** [comment]
@spapas I just mean that as workaround this middleware is fine (which is great since there is nothing else available right now), but as part of wagtail it should be done on a lower level.

**Comment by laymonage** [comment]
#10072 adds prevention of deleting pages, images, documents, and snippets that are referenced by other objects with a `ForeignKey(on_delete=models.PROTECT)` by scanning the reference index and hiding the delete confirmation button if there is at least one protected reference.

It may not be sufficient for all use cases of `PROTECT`, so if you give it a try in the next release, please give us any feedback!
